File: web/src/lib/pathways.ts
import { hypergeomSf, mean, sampleStd } from "./math";
import { log2Transform } from "./ingest";
import type { AbundanceMatrix, GeneSetLibrary, PathwayResult } from "./types";

function zscoreRows(matrix: number[][]): number[][] {
  return matrix.map((row) => {
    const mu = mean(row);
    const sd = sampleStd(row);
    if (!Number.isFinite(sd) || sd === 0) return row.map(() => Number.NaN);
    return row.map((value) => (value - mu) / sd);
  });
}

export function analyzePathways(matrix: AbundanceMatrix, geneSets: GeneSetLibrary): PathwayResult {
  const logMat = zscoreRows(log2Transform(matrix.intensities));
  const universe = new Set(matrix.proteins);
  const pathwayNames = Object.keys(geneSets);
  const sampleScores: Record<string, number[]> = {};

  for (const name of pathwayNames) {
    const genes = geneSets[name].genes.map((gene) => gene.toUpperCase()).filter((gene) => universe.has(gene));
    if (!genes.length) {
      sampleScores[name] = matrix.samples.map(() => Number.NaN);
      continue;
    }
    const idxs = genes.map((gene) => matrix.proteins.indexOf(gene));
    sampleScores[name] = matrix.samples.map((_, col) => mean(idxs.map((idx) => logMat[idx][col])));
  }

  const lots = [...new Set(matrix.lots)];
  const lotScores: Record<string, Record<string, number | null>> = {};
  for (const lot of lots) {
    lotScores[lot] = {};
    const cols = matrix.lots.map((value, idx) => (value === lot ? idx : -1)).filter((idx) => idx >= 0);
    for (const name of pathwayNames) {
      const values = cols.map((idx) => sampleScores[name][idx]);
      const scored = mean(values);
      lotScores[lot][name] = Number.isFinite(scored) ? scored : null;
    }
  }

  const means = pathwayNames.map((name) => ({
    name,
    value: mean(lots.map((lot) => lotScores[lot][name] ?? Number.NaN)),
  }));
  means.sort((a, b) => (b.value || 0) - (a.value || 0));

  const abundance = matrix.proteins.map((protein, idx) => ({
    protein,
    value: mean(log2Transform(matrix.intensities)[idx]),
  }));
  abundance.sort((a, b) => b.value - a.value);
  const querySize = Math.max(8, Math.ceil(abundance.length * 0.25));
  const query = new Set(abundance.slice(0, querySize).map((row) => row.protein));
  const nUniverse = universe.size;

  const ora = pathwayNames.map((name) => {
    const spec = geneSets[name];
    const geneset = new Set(spec.genes.map((gene) => gene.toUpperCase()).filter((gene) => universe.has(gene)));
    const overlap = [...query].filter((gene) => geneset.has(gene)).sort();
    const k = overlap.length;
    const kSet = geneset.size;
    const pValue = kSet && k ? hypergeomSf(k, nUniverse, kSet, querySize) : 1;
    return {
      pathway: name,
      description: spec.description ?? "",
      overlap: k,
      setInMatrix: kSet,
      querySize,
      pValue,
      overlapGenes: overlap.slice(0, 12).join(", "),
    };
  });
  ora.sort((a, b) => a.pValue - b.pValue);

  return {
    lotScores,
    pathwayNames,
    ora,
    topPathways: means.map((row) => row.name),
  };
}

File: secretomeqc/pathways.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

import numpy as np
import pandas as pd
from scipy.stats import hypergeom

from .config import GENE_SETS_PATH
from .ingest import AbundanceMatrix, log2_transform


@dataclass
class PathwayResult:
    sample_scores: pd.DataFrame
    lot_scores: pd.DataFrame
    ora: pd.DataFrame
    top_pathways: list[str]


def load_gene_sets(path: Path | None = None) -> dict[str, dict]:
    return json.loads((path or GENE_SETS_PATH).read_text())


def _zscore_matrix(matrix: pd.DataFrame) -> pd.DataFrame:
    mean = matrix.mean(axis=1)
    std = matrix.std(axis=1, ddof=1).replace(0, np.nan)
    return matrix.sub(mean, axis=0).div(std, axis=0)


def signature_scores(data: AbundanceMatrix, gene_sets: dict[str, dict]) -> tuple[pd.DataFrame, pd.DataFrame]:
    log_mat = _zscore_matrix(log2_transform(data.intensities))
    universe = set(log_mat.index)
    sample_cols = {}
    for name, spec in gene_sets.items():
        genes = [g.upper() for g in spec["genes"] if g.upper() in universe]
        if not genes:
            sample_cols[name] = pd.Series(np.nan, index=log_mat.columns)
        else:
            sample_cols[name] = log_mat.loc[genes].mean(axis=0)
    sample_scores = pd.DataFrame(sample_cols)
    sample_scores.index.name = "sample"
    lot = data.sample_meta.loc[sample_scores.index, "lot"]
    lot_scores = sample_scores.groupby(lot).mean()
    lot_scores.index.name = "lot"
    return sample_scores, lot_scores


def over_representation(data: AbundanceMatrix, gene_sets: dict[str, dict], top_frac: float = 0.25) -> pd.DataFrame:
    log_mat = log2_transform(data.intensities)
    abundance = log_mat.mean(axis=1).sort_values(ascending=False)
    n_query = max(8, int(np.ceil(len(abundance) * top_frac)))
    query = set(abundance.head(n_query).index)
    universe = set(abundance.index)
    n_universe = len(universe)
    rows = []
    for name, spec in gene_sets.items():
        geneset = {g.upper() for g in spec["genes"]} & universe
        overlap = sorted(query & geneset)
        k = len(overlap)
        k_set = len(geneset)
        # P(X >= k)
        pval = float(hypergeom.sf(k - 1, n_universe, k_set, n_query)) if k_set and k else 1.0
        rows.append(
            {
                "pathway": name,
                "description": spec.get("description", ""),
                "overlap": k,
                "set_in_matrix": k_set,
                "query_size": n_query,
                "p_value": pval,
                "overlap_genes": ", ".join(overlap[:12]),
            }
        )
    ora = pd.DataFrame(rows).sort_values("p_value")
    return ora


def analyze_pathways(data: AbundanceMatrix, gene_sets: dict[str, dict] | None = None) -> PathwayResult:
    gene_sets = gene_sets or load_gene_sets()
    sample_scores, lot_scores = signature_scores(data, gene_sets)
    ora = over_representation(data, gene_sets)
    ranked = lot_scores.mean(axis=0).sort_values(ascending=False)
    return PathwayResult(
        sample_scores=sample_scores,
        lot_scores=lot_scores,
        ora=ora,
        top_pathways=list(ranked.index),
    )
